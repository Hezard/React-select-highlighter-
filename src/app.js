import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { colors } from './constants';
import { appActions, appSelectors } from './ducks';

import { Button } from './components/button';
import { Highlighter } from './components/highlighter';
import { ButtonToggleGroup } from './components/buttonToggleGroup';
import { HighlighterFilter } from './components/highlighterFilter';

import './app.scss';

export class App extends Component {
  handleReset = () => {
    this.props.resetAll();
  };

  handleTextChange = e => {
    this.props.textChange(e.target.value);
  };

  handleHighlightColorChange = color => {
    this.props.colorChange(color);
  };

  handleClear = () => {
    this.props.colorChange(null);
  };

  handleSelectionChange = selections => {
    this.props.selectionsChange(selections);
  };

  handleFilterColorChange = item => {
    this.props.filtersChange(item);
  };

  render() {
    const { text, color, selections, filters, filteredSelections } = this.props;
    const highlightColor = color && color.value;

    return (
      <div className="app">
        <header className="app-header">
          <h1>React Select Highlighter</h1>
        </header>
        <section className="app-content">
          <textarea
            className="form-textfield mb-3"
            placeholder="Enter text"
            value={text}
            onChange={this.handleTextChange}
          />
          <div className="row mb-3">
            <div className="col">
              <ButtonToggleGroup
                dataTestId="btn-group-selector"
                data={colors}
                active={color}
                onChange={this.handleHighlightColorChange}
              />
            </div>
            <div className="col d-flex">
              <Button
                type="button"
                className="btn btn--outline btn--primary ml-auto"
                onClick={this.handleClear}
              >
                Eraser
              </Button>
              <Button
                type="button"
                className="btn btn--outline btn--secondary ml-2"
                onClick={this.handleReset}
                data-testid="btn-reset"
              >
                Reset
              </Button>
            </div>
          </div>
          <Highlighter
            color={highlightColor}
            text={text}
            selections={selections}
            onChange={this.handleSelectionChange}
          />
          <h4>Filters:</h4>
          <ButtonToggleGroup
            dataTestId="btn-group-filter"
            multiple
            data={colors}
            className="mb-3"
            btnClassName="btn--outline"
            active={filters}
            onChange={this.handleFilterColorChange}
          />
          <HighlighterFilter data={filteredSelections} />
        </section>
      </div>
    );
  }
}

App.propTypes = {
  color: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    value: PropTypes.string
  }),
  filters: PropTypes.shape({
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    value: PropTypes.string
  }),
  filteredSelections: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      start: PropTypes.number,
      end: PropTypes.number,
      value: PropTypes.string,
      variant: PropTypes.string
    })
  ),
  text: PropTypes.string,
  selections: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      start: PropTypes.number,
      end: PropTypes.number,
      value: PropTypes.string,
      variant: PropTypes.string
    })
  )
};

export default connect(
  state => ({
    color: appSelectors.getColor(state),
    text: appSelectors.getText(state),
    filters: appSelectors.getFilters(state),
    selections: appSelectors.getSelections(state),
    filteredSelections: appSelectors.getFilteredSelections(state)
  }),
  appActions
)(App);
