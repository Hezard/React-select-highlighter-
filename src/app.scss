@use "/scss/vars" as *;
@use "/scss/grid";

.app {
  display: flex;
  flex-direction: column;
  height: 100%;

  &-header {
    position: sticky;
    top: 0;
    z-index: 1;
    flex: none;
    padding: 0.5rem 2rem;
    border-bottom: 1px solid $gray;
    background-color: $white;
    text-align: center;
  }

  &-content {
    padding: 2rem;
    flex: 0 1 auto;

    @media (min-width: 768px) {
      margin: 0 auto;
      width: 50rem;
    }
  }
}

.form-textfield {
  border-radius: 3px;
  border: 1px solid $gray;
  padding: 1.5rem;
  width: 100%;
  min-height: 12rem;

  resize: vertical;
}
