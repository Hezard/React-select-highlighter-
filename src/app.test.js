import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { App } from './app';

describe('App integration test', () => {
  let container, getByPlaceholderText, getByText, getByTestId;
  const colorChange = jest.fn(),
    textChange = jest.fn(),
    filtersChange = jest.fn(),
    selectionsChange = jest.fn(),
    resetAll = jest.fn();
  const props = {
    color: null,
    filters: {},
    selections: [],
    text: 'defaultText',
    colorChange,
    textChange,
    filtersChange,
    selectionsChange,
    resetAll
  };

  beforeEach(() => {
    ({ container, getByPlaceholderText, getByText, getByTestId } = render(<App {...props} />));
  });

  afterEach(() => {
    jest.clearAllMocks();
    cleanup();
  });

  test('test initial render', () => {
    expect(container.firstChild).toMatchSnapshot();
  });

  test('test textarea', () => {
    const textarea = getByPlaceholderText(/Enter text/);
    const result = 'Test text';
    fireEvent.change(textarea, { target: { value: 'Test text' } });
    expect(textChange.mock.calls[0][0]).toBe(result);
  });

  test('test eraser button', () => {
    const btn = getByText(/Eraser/);
    fireEvent.click(btn);
    expect(colorChange.mock.calls[0][0]).toBe(null);
  });

  test('test reset button', () => {
    const btn = getByText(/Reset/);
    fireEvent.click(btn);
    expect(resetAll).toHaveBeenCalledTimes(1);
  });

  test('test color button group', () => {
    const btnGroup = getByTestId('btn-group-selector');

    fireEvent.click(btnGroup.firstChild);
    expect(colorChange).toHaveBeenCalledTimes(1);
    expect(colorChange.mock.calls[0][0]).toMatchObject({
      id: 0,
      value: 'red'
    });
  });

  test('test filter button group', () => {
    const btnGroup = getByTestId('btn-group-filter');

    fireEvent.click(btnGroup.firstChild);
    expect(filtersChange).toHaveBeenCalledTimes(1);
    expect(filtersChange.mock.calls[0][0]).toMatchObject({
      id: 0,
      value: 'red'
    });
  });
});
