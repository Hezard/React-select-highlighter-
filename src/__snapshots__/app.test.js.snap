// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App integration test test initial render 1`] = `
<div
  class="app"
>
  <header
    class="app-header"
  >
    <h1>
      React Select Highlighter
    </h1>
  </header>
  <section
    class="app-content"
  >
    <textarea
      class="form-textfield mb-3"
      placeholder="Enter text"
    >
      defaultText
    </textarea>
    <div
      class="row mb-3"
    >
      <div
        class="col"
      >
        <div
          class="btn-toggle-group"
          data-testid="btn-group-selector"
        >
          <button
            class="btn btn btn--primary btn--md btn--primary"
          >
            red
          </button>
          <button
            class="btn btn btn--primary btn--md btn--primary"
          >
            green
          </button>
          <button
            class="btn btn btn--primary btn--md btn--primary"
          >
            yellow
          </button>
        </div>
      </div>
      <div
        class="col d-flex"
      >
        <button
          class="btn btn btn--outline btn--primary ml-auto btn--md btn--primary"
        >
          Eraser
        </button>
        <button
          class="btn btn btn--outline btn--secondary ml-2 btn--md btn--primary"
        >
          Reset
        </button>
      </div>
    </div>
    <article
      class="highlighter"
    >
      defaultText
    </article>
    <h4>
      Filters:
    </h4>
    <div
      class="btn-toggle-group mb-3"
      data-testid="btn-group-filter"
    >
      <button
        class="btn btn btn--outline btn--md btn--primary"
      >
        red
      </button>
      <button
        class="btn btn btn--outline btn--md btn--primary"
      >
        green
      </button>
      <button
        class="btn btn btn--outline btn--md btn--primary"
      >
        yellow
      </button>
    </div>
  </section>
</div>
`;
